### PR TITLE
Fungible token payments and fiat price conversion (dw-66)

### PR DESCRIPTION
**Pull request by intrprtr**
This PR adds `FungibleConversionProxy`, a contract that allows users to pay a request denominated in fiat, with fungible tokens. Please see the associated issue #5 for more information.

### CODE REVIEW

File: fungible_conversion_proxy/Cargo.toml
Comment (yomarion) [inline fungible_conversion_proxy/Cargo.toml]:
```suggestion
authors = ["Request Network Foundation"]
```
Just a detail but since the bounty is subsidised by the foundation, it's more relevant like this.

File: fungible_conversion_proxy/src/lib.rs
Comment (yomarion) [inline fungible_conversion_proxy/src/lib.rs:152]: Shouldn't we make it `#[private]`?

File: fungible_conversion_proxy/src/lib.rs
Comment (yomarion) [inline fungible_conversion_proxy/src/lib.rs]:
```suggestion
```
Isn't payable for paying the method with NEAR tokens?

File: fungible_conversion_proxy/src/lib.rs
Comment (yomarion) [inline fungible_conversion_proxy/src/lib.rs]:
```suggestion
```
Same here and for `rate_callback`, shouldn't we skip the payable attribute for ft payments? Maybe I miss something about Near, and then we can maybe skip `env::attached_deposit()`?

File: tests/sim/main.rs
Comment (yomarion) [inline tests/sim/main.rs]:
```suggestion
const PROXY_ID: &str = "request_proxy";
const FUNGIBLE_PROXY_ID: &str = "fungible_request_proxy";
```
To avoid the confusion with fungible token contract ID.

**Review by yomarion — COMMENTED** [review]
Shouldn't you test `ft_transfer_call` in the simulated tests? It's the method everyone should call, right?

File: fungible_conversion_proxy/src/lib.rs
Comment (intrprtr) [reply fungible_conversion_proxy/src/lib.rs]: My understanding of the NEP-141 standard is that exactly 1 yoctoNEAR needs to be attached to a transfer function call for security purposes (see the "Requirements" comment above function signatures in the [spec reference](https://nomicon.io/Standards/Tokens/FungibleToken/Core#reference-level-explanation)).

File: fungible_conversion_proxy/src/lib.rs
Comment (intrprtr) [reply fungible_conversion_proxy/src/lib.rs]: Same as above.

**Comment by intrprtr** [comment]
> Shouldn't you test `ft_transfer_call` in the simulated tests? It's the method everyone should call, right?

It is the method everyone should call, but `ft_transfer_call` is defined in the fungible token contract, not in our contract. `ft_on_transfer` is the main entry point to our contract. How the fungible token contract implements `ft_transfer_call` is up to them as long as it makes a cross-contract call to the `ft_on_transfer` function in our contract. I think it doesn't make sense to test this as we are not trying to test a fungible token contract and can only assume it is implemented correctly as the specification defines. Right now the tests call `ft_on_transfer` in our contract directly (i.e. simulating what the external fungible token contract would be calling). Let me know if you feel otherwise or if I'm missing something.

**Comment by MantisClone** [comment]
The unit and integration tests passed.

<details>
<summary>Unit tests</summary>

```
mantisclone@mantisclone-Inspiron-7620:~/projects/near-contracts/fungible_conversion_proxy$ cargo test
   Compiling fungible_conversion_proxy v0.0.1 (/home/mantisclone/projects/near-contracts/fungible_conversion_proxy)
    Finished test [unoptimized + debuginfo] target(s) in 0.82s
     Running unittests src/lib.rs (/home/mantisclone/projects/near-contracts/target/debug/deps/fungible_conversion_proxy-273085b091c7f23f)

running 13 tests
test tests::test_get_transfer_with_reference_args ... ok
test tests::admin_oracle ... ok
test tests::admin_owner ... ok
test tests::admin_provider_no_permission - should panic ... ok
test tests::transfer_with_invalid_msg_format - should panic ... ok
test tests::admin_owner_no_permission - should panic ... ok
test tests::admin_provider ... ok
test tests::admin_oracle_no_permission - should panic ... ok
test tests::transfer_with_invalid_reference_length - should panic ... ok
test tests::transfer_with_invalid_reference_value - should panic ... ok
test tests::transfer_with_msg_missing_fields - should panic ... ok
test tests::transfer_with_reference ... ok
test tests::transfer_with_not_enough_gas - should panic ... ok

test result: ok. 13 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests fungible_conversion_proxy

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

</details>

<details>
<summary>Integration tests</summary>

```
mantisclone@mantisclone-Inspiron-7620:~/projects/near-contracts$ ./test.sh 
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
    Finished release [optimized] target(s) in 0.06s
    Finished test [unoptimized + debuginfo] target(s) in 0.11s
     Running tests/sim/main.rs (target/debug/deps/sim-bffe6b8153889bc5)

running 11 tests
test_transfer_with_invalid_parameter_length > TeraGas burnt: 2.823952938794
test test_transfer_with_invalid_reference_length ... ok
test test_transfer_with_wrong_currency ... ok
test test_transfer_zero_usd_near ... ok
test test_transfer_usd_fungible_fee_receiver_send_failed ... ok
test test_outdated_rate ... ok
test test_transfer_usd_fungible ... ok
test_transfer_usd_near ==> TeraGas burnt: 15.528291453811
diff: 0
test test_transfer_usd_near ... ok
test test_transfer_usd_fungible_not_enough ... ok
test test_zero_usd_fungible ... ok
test test_outdated_rate_fungible ... ok
test test_transfer_usd_fungible_receiver_send_failed ... ok

test result: ok. 11 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2.57s
```

</details>

File: fungible_conversion_proxy/src/lib.rs
Comment (MantisClone) [inline fungible_conversion_proxy/src/lib.rs]: I believe `near_sdk::assert_self();` is redundant with the `#[private]` macro and can be removed. Reference:
https://docs.near.org/sdk/rust/contract-interface/private-methods#when-using-callbacks

File: fungible_conversion_proxy/src/lib.rs
Comment (MantisClone) [inline fungible_conversion_proxy/src/lib.rs]: Same here as previous comment.

File: fungible_conversion_proxy/src/lib.rs
Comment (MantisClone) [inline fungible_conversion_proxy/src/lib.rs:339]: Why is the BASIC_GAS multiplied by 8 here?

File: fungible_conversion_proxy/src/lib.rs
Comment (MantisClone) [inline fungible_conversion_proxy/src/lib.rs]: Same comment here as above.

File: fungible_conversion_proxy/src/lib.rs
Comment (MantisClone) [inline fungible_conversion_proxy/src/lib.rs]: I don't think the `crypto_` prefix for `crypto_amount` and `crypto_fee_amount` don't add information.  I think they could be renamed to `amount` and `fee_amount`.

Same with `crypto_amount_args` and `crypto_fee_amount_args` below.

File: fungible_conversion_proxy/src/lib.rs
Comment (MantisClone) [reply fungible_conversion_proxy/src/lib.rs]: So since this function performs two transfers (main payment and fee payment) does that mean it requires 2 yoctoNEAR?

**Review by MantisClone — COMMENTED** [review]
I reviewed the functions. I'll review the tests more tomorrow.

File: fungible_conversion_proxy/src/lib.rs
Comment (intrprtr) [reply fungible_conversion_proxy/src/lib.rs:339]: The `rate_callback` function is where a lot of the logic occurs, including the rate conversion calculation and token transfer calls, so it needs more gas to execute. I'm not sure if there's a more elegant way of giving it enough gas than hardcoding a value like this.

File: fungible_conversion_proxy/src/lib.rs
Comment (intrprtr) [reply fungible_conversion_proxy/src/lib.rs]: I believe so, but I'm not certain on this. The docs don't specify whether the 1 yoctoNEAR is returned or consumed by the contract. Regardless, if the attached deposit is not sufficient for both the transfer calls to succeed, the caller should be refunded the entire original amount of tokens.

Comment (intrprtr) [reply fungible_conversion_proxy/src/lib.rs]: Actually, after looking into this further, I believe the user only needs to attach 1 yoctoNEAR to their call of `ft_transfer_call` _on the fungible token _contract__ (this has nothing to do with our code - it should be required as part of the fungible token standard).

As far as I can tell, the NEP-141 standard does not specify whether _at least_ or _exactly_ 1 yoctoNEAR is required and it seems like there may be some variation among fungible token contracts. Furthermore, it is also not specified whether the attached deposit is passed on when the contract calls `ft_on_transfer` in our contract, so we might not be able to rely on a consistent implementation one way or the other. Regardless, the two calls to `ft_transfer` in our contract will use our contract's own balance for the 1 yoctoNEAR deposit. Since these two calls only use 1 yoctoNEAR each per call, a simple intial deposit of some negligible value worth of NEAR into the contract will be enough to ensure it functions properly forever, so we don't have to worry about what is or isn't attached by the calling fungible token contract.

Comment (yomarion) [reply fungible_conversion_proxy/src/lib.rs]: I think that the minimum deposit applies to `ft_transfer` but not to `ft_on_transfer`, I can see an example from the Near foundation here: https://github.com/near-examples/FT/blob/334b32d65313f8c78d614e7cb4ba629af3af0fd9/test-contract-defi/src/lib.rs#L37 that is not payable.
If this is a correct analogy, none of these methods should be `payable`.

I agree that this is not related to gas saving but more for future maintainers to understand the methods.

Comment (intrprtr) [reply fungible_conversion_proxy/src/lib.rs]: I agree, this makes sense to me. I have made the change.

File: fungible_conversion_proxy/src/lib.rs
Comment (yomarion) [inline fungible_conversion_proxy/src/lib.rs]:
```suggestion
pub struct PaymentArgs {
```

File: fungible_conversion_proxy/src/lib.rs
Comment (yomarion) [inline fungible_conversion_proxy/src/lib.rs]:
```suggestion
        let main_transfer_args =
```

File: fungible_conversion_proxy/src/lib.rs
Comment (yomarion) [inline fungible_conversion_proxy/src/lib.rs]:
```suggestion
        let fee_transfer_args =
```

File: fungible_conversion_proxy/src/lib.rs
Comment (yomarion) [inline fungible_conversion_proxy/src/lib.rs]: The token address should be inferred from the predecessor account instead of `msg`. It makes the interface lighter and strengthens the contract security. args.token_address should never be different from the tokens sent. The current logic makes it unlikely but not explicitly impossible.

**Review by yomarion — COMMENTED** [review]
A couple more changes and then we are good to merge!

File: fungible_conversion_proxy/test.sh
Comment (MantisClone) [inline fungible_conversion_proxy/test.sh:2]: Context: While reviewing this PR, I realized that we didn't have automated testing so I added some in #8.

Can you please add a step to `build-and-test.yml`?

```yml
    - name: Build and run fungible conversion proxy unit tests
      working-directory: ./fungible_conversion_proxy
      run: cargo test
```

Can you please also update the `near-contracts/README.md` under "Unit tests"?

```console
cd near-contracts/fungible_conversion_proxy
cargo test
```

Comment (intrprtr) [reply fungible_conversion_proxy/test.sh:2]: Done!

File: mocks/src/fpo_oracle_mock.rs
Comment (MantisClone) [inline mocks/src/fpo_oracle_mock.rs:84]: I just realized there are mock tests too. Can you please add them to the automated tests and the README?

Comment (intrprtr) [reply mocks/src/fpo_oracle_mock.rs:84]: Done!

File: tests/sim/main.rs
Comment (MantisClone) [inline tests/sim/main.rs]:
```suggestion
const FUNGIBLE_PROXY_ID: &str = "fungible_request_proxy";
```

File: tests/sim/main.rs
Comment (MantisClone) [inline tests/sim/main.rs]:
```suggestion
// Helper function for setting up fungible token transfer tests
```

File: mocks/src/fungible_token_mock.rs
Comment (MantisClone) [inline mocks/src/fungible_token_mock.rs]: Can you please rename this to `ft_balance_of` to match the NEP-141 standard.

File: tests/sim/main.rs
Comment (MantisClone) [inline tests/sim/main.rs]: Can you please replace some of these numbers with named constants to make this calculation more self-explanatory?  I understand that `102` USDC.e was spent and I think that the first `1000000` is the USDC.e decimals, but I don't understand the `1000000 / 999900`.

Same comment for all the other calculations like this in this file.

File: tests/sim/main.rs
Comment (MantisClone) [inline tests/sim/main.rs:114]: @intrprtr I'm concerned about what you said in a previous comment:

> Since these two calls only use 1 yoctoNEAR each per call, a simple intial deposit of some negligible value worth of NEAR into the contract will be enough to ensure it functions properly forever, so we don't have to worry about what is or isn't attached by the calling fungible token contract.

I wish there was a test where the fungibleConversionProxy has 0 yocto and alice pays the 2 yocto, just to show that it's possible (in case the fungibleConversionProxy's NEAR reserves run out).

More generally, I wish there were tests where the entry point was `ft_transfer` instead of `ft_on_transfer`

File: mocks/src/fungible_token_mock.rs
Comment (MantisClone) [inline mocks/src/fungible_token_mock.rs]: Can you please add the expected panic message?

File: mocks/src/fungible_token_mock.rs
Comment (MantisClone) [inline mocks/src/fungible_token_mock.rs]: Same comment as above.

File: mocks/src/fungible_token_mock.rs
Comment (MantisClone) [inline mocks/src/fungible_token_mock.rs]: Same comment as above.

File: mocks/src/fungible_token_mock.rs
Comment (MantisClone) [inline mocks/src/fungible_token_mock.rs]: Can you please remove this blank line?
```suggestion
```

File: mocks/src/fungible_token_mock.rs
Comment (MantisClone) [inline mocks/src/fungible_token_mock.rs]: Can you please remove this blank line?
```suggestion
```

**Review by MantisClone — CHANGES_REQUESTED** [review]

File: tests/sim/main.rs
Comment (intrprtr) [reply tests/sim/main.rs:114]: > I wish there was a test where the fungibleConversionProxy has 0 yocto and alice pays the 2 yocto, just to show that it's possible (in case the fungibleConversionProxy's NEAR reserves run out).

@MantisClone As discussed [here](https://github.com/RequestNetwork/near-contracts/pull/7#discussion_r1087453540), I think we _cannot_ assume that it is possible to have alice pay the 2 yoctoNEAR. As far as I can tell, the NEP-141 standard does _not_ specify whether or not exactly one or at least one yoctoNEAR should be deposited, and does _not_ specify whether excess deposit should be passed onto the cross-contract call. Therefore, it is entirely possible to satisfy the NEP-141 standard _without_ allowing any way for the user to pay the yoctoNEAR to our contract. Since this is a possibility and we want to generally support fungible tokens, I think it makes more sense to operate under the assumption that the yoctoNEAR must come out of `fungibleConversionProxy`'s balance. `fungibleConversionProxy` will need to maintain a balance to pay for storage anyways as required for all NEAR contracts. Since 1 yoctoNEAR = 10^-24 NEAR and each call only uses 2 yoctoNEAR, even a tiny initial deposit is enough to ensure the deposit is covered forever.

> More generally, I wish there were tests where the entry point was `ft_transfer` instead of `ft_on_transfer`

I assume you mean `ft_transfer_call` instead of `ft_transfer`, as `ft_transfer` is just the simple transfer function in the NEP-141 standard (no cross-contract call, just balance changes). As we discussed [here](https://github.com/RequestNetwork/near-contracts/pull/7#issuecomment-1404165142), testing calls to `ft_transfer_call` wouldn't make sense as it's not defined in our contract - it's defined in the external fungible token contract and the implementation can vary between fungible tokens. Our only guarantee from the NEP-141 standard is that it will make a cross-contract call to `ft_on_transfer` in our contract, which is our entry point.

**Comment by intrprtr** [comment]
@MantisClone @yomarion Just wanted to flag this again as I haven't heard from either of you in a while. I have made all the requested changes and passed the checks. If there are no other comments, could you please approve the PR?

**Review by yomarion — APPROVED** [review]
Good job! Thanks a lot for your contribution

**Review by MantisClone — APPROVED** [review]
Thanks @intrprtr for your work on this!

**Pull request merged**
